runtime/src/bank.rs

```diff
@@ -3085,6 +3087,104 @@ impl Bank {
         }
     }
 
+    // TODO - solution 1, check vote redundancy by VoteState within bank,
+    //     this solution does not alight with bankless leader goal;
+    //     mainly using it to compare with solution #2 to confirm
+    //     behavior and performance.
+    fn is_redundant_vote_by_vote_state(&self, vote_account_pubkey: &Pubkey, vote: &Vote) -> bool {
+        // ignore vote without slot, or its last slot is '0'
+        let last_slot = match vote.slots.last() {
+            None => {
+                debug!(
+                    "Vote  has no slots, skip checking redundancy for vote {:?}",
+                    vote
+                );
+                return false;
+            }
+            Some(slot) => {
+                if 0 == *slot {
+                    debug!(
+                        "Vote on slot 0, skip checking redundancy for vote {:?}",
+                        vote
+                    );
+                    return false;
+                }
+                slot
+            }
+        };
+
+        // using vote_acount_pubkey from Vote TX to get vote account from bank
+        let vote_account = match self.get_vote_account(vote_account_pubkey) {
+            None => {
+                warn!(
+                    "Vote account {} does not exist, skip checking redundancy for vote {:?}",
+                    vote_account_pubkey, vote
+                );
+                return false;
+            }
+            Some((_stake, vote_account)) => vote_account,
+        };
+
+        let vote_state = vote_account.vote_state();
+        let vote_state = match vote_state.as_ref() {
+            Err(_) => {
+                warn!(
+                    "Vote account {} is unreachable, skip checking redundancy for vote {:?}",
+                    vote_account_pubkey, vote
+                );
+                return false;
+            }
+            Ok(vote_state) => vote_state,
+        };
+
+        match vote_state.check_slots_are_valid(vote, &[(*last_slot, vote.hash)]) {
+            Err(VoteError::VoteTooOld) => {
+                debug!(
+                    "Vote {:?} by vote account  {:?} is redundant",
+                    vote, vote_account_pubkey
+                );
+                true
+            }
+            _ => false,
+        }
+    }
+
+    fn check_redundant_votes(
+        &self,
+        sanitized_txs: &[SanitizedTransaction],
+        lock_results: &[Result<()>],
+    ) -> Vec<Result<()>> {
+        sanitized_txs
+            .as_transactions_iter()
+            .zip(lock_results.to_vec())
```

Comment:
I think you could pass in this zip as an iterator for more clarity, i.e. something like `txs_and_locks: impl Iterator<Item=(&SanitizedTransaction, &Result<()>)>`